plugins/MidiExport/MidiFile.hpp

```diff
@@ -0,0 +1,320 @@
+
+int writeBigEndian2(uint16_t val, uint8_t *buf)
+{
+	buf[0] = val >> 8 & 0xff;
+	buf[1] = val & 0xff;
+	return 2;
+}
+
+
+class MIDIHeader
+{
+	// Class to encapsulate the MIDI header structure.
+	uint16_t numTracks;
+	uint16_t ticksPerBeat;
+	
+	public:
+	
+	MIDIHeader(uint16_t nTracks, uint16_t ticksPB=TICKSPERBEAT): numTracks(nTracks), ticksPerBeat(ticksPB) {}
+	
+	inline int writeToBuffer(uint8_t *buffer, int start=0) const
+	{
+		// chunk ID
+		buffer[start++] = 'M'; buffer[start++] = 'T'; buffer[start++] = 'h'; buffer[start++] = 'd';
+		// chunk size (6 bytes always)
+		buffer[start++] = 0; buffer[start++] = 0; buffer[start++] = 0; buffer[start++] = 0x06;
+		// format: 1 (multitrack)
+		buffer[start++] = 0; buffer[start++] = 0x01;
+		
+		start += writeBigEndian2(numTracks, buffer+start);
+		
+		start += writeBigEndian2(ticksPerBeat, buffer+start);
+		
+		return start;
+	}
+	
+};
+
+
+struct Event
+{
+	uint32_t time;
+	uint32_t tempo;
+	string trackName;
+	enum {NOTE_ON, NOTE_OFF, TEMPO, PROG_CHANGE, TRACK_NAME} type;
+	// TODO make a union to save up space
+	uint8_t pitch;
+	uint8_t programNumber;
+	uint8_t duration;
+	uint8_t volume;
+	uint8_t channel;
+	
+	Event() {time=tempo=pitch=programNumber=duration=volume=channel=0; trackName="";}
+	
+	inline int writeToBuffer(uint8_t *buffer) const 
+	{
+		uint8_t code, fourbytes[4];
+		int size=0;
+		switch (type) 
+		{
+			case NOTE_ON:
+				code = 0x9 << 4 | channel;
+				size += writeVarLength(time, buffer+size);
+				buffer[size++] = code;
+				buffer[size++] = pitch;
+				buffer[size++] = volume;
+				break;
+			case NOTE_OFF:
+				code = 0x8 << 4 | channel;
+				size += writeVarLength(time, buffer+size);
+				buffer[size++] = code;
+				buffer[size++] = pitch;
+				buffer[size++] = volume;
+				break;
+			case TEMPO:
+				code = 0xFF;
+				size += writeVarLength(time, buffer+size);
+				buffer[size++] = code;
+				buffer[size++] = 0x51;
+				buffer[size++] = 0x03;
+				writeBigEndian4(int(60000000.0 / tempo), fourbytes);
+				
+				//printf("tempo of %x translates to ", tempo);
+				for (int i=0; i<3; i++) printf("%02x ", fourbytes[i+1]);
+				printf("\n");
+				buffer[size++] = fourbytes[1];
+				buffer[size++] = fourbytes[2];
+				buffer[size++] = fourbytes[3];
+				break;
+			case PROG_CHANGE:
+				code = 0xC << 4 | channel;
+				size += writeVarLength(time, buffer+size);
+				buffer[size++] = code;
+				buffer[size++] = programNumber;
+				break;
+			case TRACK_NAME:
+				size += writeVarLength(time, buffer+size);
+				buffer[size++] = 0xFF;
+				buffer[size++] = 0x03;
+				size += writeVarLength(trackName.size(), buffer+size);
+				trackName.copy((char *)(&buffer[size]), trackName.size());
+				size += trackName.size();
+//				 buffer[size++] = '\0';
+//				 buffer[size++] = '\0';
+				
+				break;
+		}
+		return size;
+	} // writeEventsToBuffer
+	
+	
+	// events are sorted by their time
+	inline bool operator < (const Event& b) const {
+		if (this->time < b.time) return true;
+#if 0
```

Comment:
Can you clean this up?

Remember that you don't have to create additional commits, after you've done your changes and added them you can use `git commit --amend` instead of the regular one to ammend your changes to the last commit.

You'll have to push with `-f` later. This is not an issue since this branch is only yours.